### PR TITLE
Sign Bundle with a Timestamp Authority

### PR DESCRIPTION
**Pull request by DarkaMaul**
Final bit of https://github.com/sigstore/sigstore-python/issues/1182

#### Summary

This PR introduces the possibility to create a bundle with a timestamp signed by a TimestampAuthority.

### Added

- If a `Timestamp Authority` URL has been provided in the SigningConfig, the bundle are now automatically generated with a signed timestamp.  

/cc @woodruffw @facutuesca

### CODE REVIEW

File: sigstore/_internal/timestamp.py [hunk 1]
```diff
@@ -20,6 +20,17 @@
...
+import requests
+from rfc3161_client import (
+    TimestampRequestBuilder,
+    TimeStampResponse,
+    decode_timestamp_response,
+)
+
+from sigstore._internal import USER_AGENT
+
+CLIENT_TIMEOUT: int = 30
```
Comment (woodruffw) [inline sigstore/_internal/timestamp.py:32]: This seems like a really long timeout. How was this picked, and can we just use the default timeout instead?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -58,6 +58,18 @@ jobs:
...
         run: make test TEST_ARGS="-vv --showlocals"
 
+      - name: test (timestamp-authority)
+        if: ${{ matrix.conf.os == 'ubuntu-latest' }}
+        run: |
+          wget https://github.com/sigstore/timestamp-authority/releases/download/${SIGSTORE_TIMESTAMP}/timestamp-server-linux-amd64 -O /tmp/timestamp-server
+          chmod +x /tmp/timestamp-server
+          # Run the TSA in background
+          /tmp/timestamp-server serve --port 3000 --disable-ntp-monitoring &
+          export SIGSTORE_TIMESTAMP_AUTHORITY_URL="http://localhost:3000/api/v1/timestamp"
```
Comment (woodruffw) [inline .github/workflows/ci.yml:68]:
```suggestion
          export TEST_SIGSTORE_TIMESTAMP_AUTHORITY_URL="http://localhost:3000/api/v1/timestamp"
```

File: sigstore/_internal/timestamp.py (hunk 1 above)
Comment (DarkaMaul) [reply sigstore/_internal/timestamp.py:32]: I used `sigstore-go` [value](https://github.com/sigstore/sigstore-go/blob/e92142f0734064ebf6001f188b7330a1212245fe/pkg/sign/timestamping.go#L57)

**Comment by woodruffw** [comment]
This looks like something bad on the Windows CI:

```
    Error configuring OpenSSL build:
        'perl' reported failure with exit code: 2
        Command failed: "perl" "./Configure" "--prefix=C:/Users/runneradmin/AppData/Local/Temp/pip-install-ldkkdr4i/rfc3161-client_a9d25b6f62a14f8aac4a34a83ab37a09/rust/target/release/build/openssl-sys-b380da04be89fe31/out/openssl-build/install" "--openssldir=SYS$MANAGER:[OPENSSL]" "no-shared" "no-ssl3" "no-tests" "no-comp" "no-zlib" "no-zlib-dynamic" "--libdir=lib" "no-md2" "no-rc5" "no-weak-ssl-ciphers" "no-camellia" "no-idea" "no-seed" "no-capieng" "no-asm" "VC-WIN64A"
  
```

**Comment by jku** [comment]
> This looks like something bad on the Windows CI

I think this might require some of the same build config tricks that rfc3161-client CI has... alternatively maybe wait for a  rfc161-client release and avoid installing from git

**Comment by woodruffw** [comment]
> > This looks like something bad on the Windows CI
> 
> I think this might require some of the same build config tricks that rfc3161-client CI has... alternatively maybe wait for a rfc3161-client release and avoid installing from git

Yeah, I think we can prep another release. CC @DarkaMaul

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -58,6 +58,19 @@ jobs:
...
+      - name: test (timestamp-authority)
+        if: ${{ matrix.conf.os == 'ubuntu-latest' }}
+        run: |
+          wget https://github.com/sigstore/timestamp-authority/releases/download/${SIGSTORE_TIMESTAMP}/timestamp-server-linux-amd64 -O /tmp/timestamp-server
+          chmod +x /tmp/timestamp-server
+          # Run the TSA in background
+          /tmp/timestamp-server serve --port 3000 --disable-ntp-monitoring &
+          export TEST_SIGSTORE_TIMESTAMP_AUTHORITY_URL="http://localhost:3000/api/v1/timestamp"
+          make test TEST_ARGS="-m timestamp_authority -rs" | tee output
+          ! grep -q "skipping test that requires a Timestamp Authority" output || (echo "ERROR: Found skip message" && exit 1)
```
Comment (woodruffw) [inline .github/workflows/ci.yml:70]: This line could use a comment -- it looks load bearing, but I'm having trouble understanding what it does.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -58,6 +58,21 @@ jobs:
...
+          chmod +x /tmp/timestamp-server
+          # Run the TSA in background
+          /tmp/timestamp-server serve --port 3000 --disable-ntp-monitoring &
+          export TEST_SIGSTORE_TIMESTAMP_AUTHORITY_URL="http://localhost:3000/api/v1/timestamp"
+          # Ensure Timestamp Authority tests are not skipped by
+          # having pytest show skipped tests and verifying ours are running
+          make test TEST_ARGS="-m timestamp_authority -rs" | tee output
+          ! grep -q "skipping test that requires a Timestamp Authority" output || (echo "ERROR: Found skip message" && exit 1)
+        env:
+          SIGSTORE_TIMESTAMP: "v1.2.3"
```
Comment (woodruffw) [inline .github/workflows/ci.yml:74]: I don't _love_ that we pull down a binary here, and hardcode the version. It'd be great if we could get this into Dependabot somehow.

(Not a blocker, flagging as a follow-up.)

File: sigstore/sign.py [hunk 5]
```diff
@@ -296,7 +307,12 @@ class SigningContext:
     """
 
     def __init__(
-        self, *, fulcio: FulcioClient, rekor: RekorClient, trusted_root: TrustedRoot
+        self,
+        *,
+        fulcio: FulcioClient,
+        rekor: RekorClient,
+        trusted_root: TrustedRoot,
+        tsa_clients: List[TimestampAuthorityClient] | None = None,
```
Comment (woodruffw) [inline sigstore/sign.py:315]: Flagging: it's not ideal that this parameter list continues to grow; I think we could probably slice it down substantially by passing the entire `ClientTrustConfig` and doing the instantiations internally.

That'd be good for a follow-on refactor PR here.

**Comment by woodruffw** [comment]
This also needs a CHANGELOG entry, but I'm going to merge as-is and do a follow-up for that.

**Pull request merged**